src/python/zquantum/optimizers/scipy_optimizer.py

```diff
@@ -40,7 +40,8 @@ def default_callback(params):
                                         initial_params,
                                         method=self.method,
                                         options=self.options,
-                                        callback=callback)
+                                        callback=callback,
+                                        jac=cost_function.get_gradient)
```

Comment:
I'm not sure whether we want to do that by default.
Do all the Scipy optimizers accept `jac` as input param?
Don't we want to put a flag indicating that to `options`?
I know that for L-BFGS-B if no gradient is provided, scipy uses its own finite gradients implementation. Our implementation might be less efficient than theirs, so I'd rather not override it by default.
What do you think @simonwa7 ?

Reply:
I'm pretty sure not all scipy optimizers use `jac` (ex: COBYLA) but it does not break them to give the option. 
Also I'm pretty sure scipy finite difference implementation is nothing special and just calls the cost function twice for each gradient component, making some assumptions about the necessary step size that may not always be true, but it may be worth checking.

Reply:
I agree with @jgonthier . I've tested locally that it doesn't break the optimizers that don't require it. 

The default jacobian implementation from Scipy may be "more efficient", but it still uses the same number of calls to the `CostFunction`, so I don't see it as inherently different from our implementation. I would think that if we did think that Scipy's implementation was superior, then we would want to refactor our `finite_difference` implementation in the `CostFunction` interface to use it instead of adding more flags here.

Reply:
Sounds good then!